options/posix/generic/pthread-stubs.cpp

```diff
@@ -136,8 +136,59 @@ int pthread_equal(pthread_t t1, pthread_t t2) {
 	return 0;
 }
 
+namespace {
+	struct key_global_info {
+		bool in_use;
+
+		void (*dtor)(void *);
+		uint64_t generation;
+	};
+
+	struct key_local_info {
+		void *value;
+		uint64_t generation;
+	};
+
+	constinit frg::array<
+		key_global_info,
+		PTHREAD_KEYS_MAX
+	> key_globals_{};
+
+	thread_local constinit frg::array<
+		key_local_info,
+		PTHREAD_KEYS_MAX
+	> key_locals_{};
+
+	FutexLock key_mutex_;
+}
+
 int pthread_exit(void *ret_val) {
-	auto self = static_cast<Tcb *>(mlibc::get_current_tcb());
+	auto self = mlibc::get_current_tcb();
+
+	auto hand = self->cleanupEnd;
+	while (hand) {
+		auto old = hand;
+		hand->func(hand->arg);
+		hand = hand->prev;
+		frg::destruct(getAllocator(), old);
+	}
+
+	// FIXME: This is a hack used to avoid doing syscalls when being cancelled, as that
```

Comment:
This can be fixed by setting the tcbExitingBit when beginning the exit procedure.